File: Python/union_find.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])  # path compression
        return self.parent[x]

    def union(self, x, y):
        root_x = self.find(x)
        root_y = self.find(y)

        if root_x == root_y:
            return False  # already connected

        # union by rank
        if self.rank[root_x] < self.rank[root_y]:
            self.parent[root_x] = root_y
        elif self.rank[root_x] > self.rank[root_y]:
            self.parent[root_y] = root_x
        else:
            self.parent[root_y] = root_x
            self.rank[root_x] += 1

        return True

# Example usage
uf = UnionFind(5)
print(uf.union(0, 1))  # True
print(uf.union(1, 2))  # True
print(uf.union(0, 2))  # False (already connected)
print(uf.union(3, 4))  # True
print(uf.find(0))  # 0
print(uf.find(1))  # 0
print(uf.find(2))  # 0
print(uf.find(3))  # 3
print(uf.find(4))  # 3

File: TypeScript/union_find.ts
class UnionFind2 {
    private parent: number[];
    private rank: number[];

    constructor(n: number) {
        // Create an array [0, 1, 2, ..., n-1]
        this.parent = Array.from({ length: n }, (_, i) => i);
        this.rank = new Array(n).fill(0);
    }

    find(x: number): number {
        if (this.parent[x] !== x) {
            // Path compression: recursively find the root and update parent
            this.parent[x] = this.find(this.parent[x]);
        }
        return this.parent[x];
    }

    union(x: number, y: number): boolean {
        const rootX = this.find(x);
        const rootY = this.find(y);

        if (rootX === rootY) {
            return false; // Already connected
        }

        // Union by rank: attach the smaller tree under the larger tree
        if (this.rank[rootX] < this.rank[rootY]) {
            this.parent[rootX] = rootY;
        } else if (this.rank[rootX] > this.rank[rootY]) {
            this.parent[rootY] = rootX;
        } else {
            this.parent[rootY] = rootX;
            this.rank[rootX] += 1;
        }

        return true;
    }
}

// Example usage
const uf = new UnionFind2(5);
console.log(uf.union(0, 1)); // true
console.log(uf.union(1, 2)); // true
console.log(uf.union(0, 2)); // false (already connected)
console.log(uf.union(3, 4)); // true
console.log(uf.find(0));     // 0
console.log(uf.find(1));     // 0
console.log(uf.find(2));     // 0
console.log(uf.find(3));     // 3
console.log(uf.find(4));     // 3
